include/net/sch_generic.h

```diff
@@ -742,6 +742,44 @@ static inline struct sk_buff *qdisc_peek_head(struct Qdisc *sch)
 	return qh->head;
 }
 
+static inline struct sk_buff *__qdisc_dequeue_tail(struct qdisc_skb_head *qh)
```

Comment:
The function `__qdisc_dequeue_tail()` already exists in the kernel: https://elixir.bootlin.com/linux/v4.5/source/include/net/sch_generic.h#L644

Moreover, the kernel version of `__qdisc_dequeue_tail()` works on `struct sk_buff_head` instead of on `struct qdisc_skb_head`, what constitutes further support to replace our use of `struct qdisc_skb_head` in favor of `struct sk_buff_head` in `struct gkprio_sched_data`. See my comment at `struct gkprio_sched_data` for more information.

Reply:
Hi Michel, Im working on the changes, and one thing I came across: the current linux and net-next repo: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/sch_generic.h
do not have the`__qdisc_dequeue_tail()` function... they removed it in one of the v4.8 rcs.  Being the case, I'm not completely sure if we should incorporate the function in the patch (after it being removed once already), any thoughts on this?

Reply:
Hi @ndev2,

I've found the exact commit that drops `__qdisc_dequeue_tail()`: `a09ceb0e08140a1eec05b49b4c232d3481339cb0`. You can see what it does with the following command: `git diff -r a09ceb0e08140a1eec05b49b4c232d3481339cb0~ -r a09ceb0e08140a1eec05b49b4c232d3481339cb0`. Essentially, it dopped `__qdisc_dequeue_tail()` because after the changes the commits makes, there were no longer callers to `__qdisc_dequeue_tail()`.

Nevertheless, the function `__skb_dequeue_tail()`, which `__qdisc_dequeue_tail()` has used, is still available in the kernel. Thus, you don't need to reintroduce `__qdisc_dequeue_tail()`, but can take advantage of the previous code of `__qdisc_dequeue_tail()` to implement the code we need.

Reply:
Right ok. But if we're incorporating this function again, should'nt we try to make it parallel to `__qdisc_dequeue_head()`? and that function uses `qdisc_skb_head` as the parameter. My thinking in using `qdisc_skb_head` instead of `sk_buff_head` is that they've provided API just for our cases like ours, and making use of it would be better practise, hope that's ok. Let me know what you think.

Reply:
I don't know why they have decided to make `skb->prev` undefined with `struct qdisc_skb_head`, but without `skb->prev`, we can't have a function that drops at the tail. That's why we need to go with `struct sk_buff_head`.

Given that the code of `__qdisc_dequeue_head()` is small and we only need it once in `gkprio_dequeue()`, we can just use the code of `__qdisc_dequeue_head()` directly in `gkprio_dequeue()`. This would make the patch simpler for the kernel reviewers.

Reply:
Hi Michel, skb->prev is still defined, it is part of the `struct sk_buff` construct and not part of the list construct. I think what they mean is that the reference to the previous `sk_buff` which appeared before the start of the `sk_buff_head` list is no longer maintained (`sk_buff_head` is written with an idea to represent and maintain multiple `sk_buff` lists with operations to support splicing, joining of lists, etc. While `qdisc_skb_head` is written to represent a single queue of `sk_buff`, please take a look at these links to:  https://elixir.bootlin.com/linux/v4.16-rc5/source/include/linux/skbuff.h    https://elixir.bootlin.com/linux/v4.16-rc5/source/include/net/sch_generic.h )  Our code in `gkprio_enqueue()` is currently able to drop packets at the tail of the list.